ArduCopter/mode_guided.cpp

```diff
@@ -374,8 +378,13 @@ void ModeGuided::takeoff_run()
         copter.landinggear.retract_after_takeoff();
 
         // switch to position control mode but maintain current target
-        const Vector3f& target = wp_nav->get_wp_destination();
-        set_destination(target);
+        Location dest_loc(wp_nav->get_wp_destination());
+        Location::AltFrame frame = Location::AltFrame::ABOVE_HOME;
+        if (wp_nav->rangefinder_used_and_healthy() && wp_nav->get_terrain_source() == AC_WPNav::TerrainSource::TERRAIN_FROM_RANGEFINDER) {
```

Comment:
I worry a bit that re-doing this decision making could lead to a different decision than was originally made at takeoff.  For example if the rangefinder became healthy during the takeoff the decision could change.  This could happen if the rangefinder was out-of-range-low while on the ground. I think a simpler approach would be:

- call set_destination(target) and fill in all the arguments including the final "terrain_alt" using AC_WPNav::origin_and_destination_are_terrain_alt().
- call Mode:Guided::set_destination(Location) and provide it the location from AC_WPNav::get_wp_destination(Location& destination).

I'd probably recommend the first option.

Reply:
Thank you for the review.
I overlooked the changes during takeoff.

Reply:
Sorry to be annoying about this but could we use the 1st approach which involves calling AC_WPNav::origin_and_destination_are_terrain_alt()?  The complexity and number of lines should be much reduced.. it should shrink down to just two lines I think.

Reply:
I've tried to change the following.
```
dest_loc.set_alt_cm(dest_loc.alt, wp_nav->origin_and_destination_are_terrain_alt() ? Location::AltFrame::ABOVE_TERRAIN : Location::AltFrame::ABOVE_HOME);
set_destination(dest_loc);
```

The target altitude seems to change a little when using AVOBE_HOME.
I tested like this.

mode guided
arm throttle
takeoff 10

dest_loc.alt = 985
dest_loc.get_alt_cm(Location::AltFrame::ABOVE_HOME, target_alt_cm) = 1000

I think that I have to use get_alt_cm.
This error increases to nearly 1m when using GND_ALT_OFFSET.